handler.go

```diff
@@ -168,14 +167,36 @@ func (s *RPCServer) handleReader(ctx context.Context, r io.Reader, w io.Writer,
 				s.maxRequestSize))
 		return
 	}
+	// check if bufferedRequest is array or not.
+	if bufferedRequest.Bytes()[0] != '[' {
+		if err := json.Unmarshal(bufferedRequest.Bytes(), &req); err != nil {
+			rpcError(wf, &req, rpcParseError, xerrors.Errorf("unmarshaling request: %w", err))
+			return
+		}
+		s.doHandleRequest(ctx, wf, req, *bufferedRequest, rpcError)
+
+	} else {
+		var reqs []request
+		if err := json.Unmarshal(bufferedRequest.Bytes(), &reqs); err != nil {
+			rpcError(wf, &req, rpcParseError, xerrors.Errorf("unmarshaling requests: %w", err))
+			return
+		}
+		for _, req := range reqs {
+			s.doHandleRequest(ctx, wf, req, *bufferedRequest, rpcError)
```

Comment:
Will a single failing request make the whole batch fail? Pretty sure the JSON-RPC 2.0 doesn't expect atomicity here, but rather individiual reporting of failures. We need to test a variety of failure scenarios:

1. Two out of four requests carry invalid ID.
2. Two out of four requests fail functionally (generate an error in their handlers).
3. Two out of four requests carry the same ID.
4. Two out of four requests fail to deserialize.

I suspect that in all these cases, the JSON-RPC 2.0 spec would mandate that those failures don't prevent the other two successful requests from being processed successfully.